### PR TITLE
Gmadges/536 pointcloud example

### PR DESCRIPTION
**Pull request by Gmadges**
Added a center & bounds recalculation.
Added support for displaying color attributes.
added a new example to show off the colours a bit.

Screenshot 2020-03-01 at 11 01 47

### CODE REVIEW

File: examples/deck.gl/pointcloud/app.js [hunk 1]
```diff
@@ -81,30 +82,54 @@ export default class App extends PureComponent {
...
       pointsCount: null,
-      points: null,
+      pointsAttribute: null,
+      colorsAttribute: null,
       loadTimeMs: undefined
     });
     load(uri).then(this._onLoad.bind(this));
   }
 
+  _getBounds(attributes) {
```
Comment (ibgreen) [inline examples/deck.gl/pointcloud/app.js]: This is great! 

As a small nit, methods that don't access `this` (pure helper functions calculating output based on their input with no side effects) are usually best structured as global helper functions.

--- 

Not for this PR: For the long term: 
- I feel applications should not need to do this
- (too much coding required just to show a simple file - we really want this example to be short and simple to show off how easy it is to use our point cloud loaders). 
- Bounds calculation should happen automatically, either inside the point cloud loaders, or deck.gl (or both).

File: examples/deck.gl/pointcloud/app.js [hunk 2]
```diff
@@ -113,25 +138,31 @@ export default class App extends PureComponent {
         // in which case the vertex count is not correct for display as points
         // Proposal: Consider adding a `mesh.points` or `mesh.pointcloud` option to mesh loaders
         // in which case the loader throws away indices and just return the vertices?
-        pointsCount: attributes.POSITION.value.length / 3, // header.vertexCount,
-        points: attributes.POSITION.value,
+        pointsCount: attributes.POSITION.value.length / attributes.POSITION.size,
```
Comment (ibgreen) [inline examples/deck.gl/pointcloud/app.js]: This is fine, though my preference would be the the application doesn't have to destructure the loaded data and store it in `pieces` like this.

Maybe just call the proposed `convertMeshToDeckPointCloudData` function here and store the `data` result?

File: examples/deck.gl/pointcloud/app.js [hunk 3]
```diff
@@ -113,25 +138,31 @@ export default class App extends PureComponent {
...
-      points &&
+      pointsAttribute &&
         new PointCloudLayer({
           // Layers can't reinitialize with new binary data
           id: `point-cloud-layer-${selectedExample}`,
           coordinateSystem: COORDINATE_SYSTEM.IDENTITY,
           numInstances: pointsCount,
-          instancePositions: points,
+          data: {
+            attributes: {
```
Comment (ibgreen) [inline examples/deck.gl/pointcloud/app.js]: This is where you could pass in the output of the `convertToDeckPointCloudData` function, and possibly drop `numInstance`.

File: examples/deck.gl/pointcloud/file-index.js [hunk 4]
```diff
@@ -14,6 +14,9 @@ export default {
     },
     'Bun Zipper (Text)': {
       uri: `${LOADERS_URI}/modules/ply/test/data/bun_zipper.ply`
+    },
+    'Richmond Azaelias': {
```
Comment (ibgreen) [inline examples/deck.gl/pointcloud/file-index.js]: That is a beautiful cloud!

Note that each module's `test/data` folder has a README that lists the licenses of any data files we copy into this repository. Just make sure this file is under a permissive license and that this is stated in that README, to make sure there are no licensing concerns with merging this PR.

Comment (Gmadges) [reply examples/deck.gl/pointcloud/file-index.js]: very good point! I've added it now

File: examples/deck.gl/pointcloud/app.js (hunk 3 above)
Comment (Gmadges) [reply examples/deck.gl/pointcloud/app.js]: yep, I realised that the data object can have a length attribute which should accomplish the same thing.

File: examples/deck.gl/pointcloud/app.js (hunk 2 above)
Comment (Gmadges) [reply examples/deck.gl/pointcloud/app.js]: very good idea I've implemented this now in my latest commit

**Comment by ibgreen** [comment]
@Gmadges This looks fantastic! As soon as you sign the CLA we will merge.

Just as an idea, the next step could be to add bounds calculation to all the mesh category loaders (there are 5 of them I believe), then we could remove that clutter from the example.

We could just add your function here to each loader, or we could place a helper in `loader-utils` module for now, or we could create `meshes` "category module" for such utilities.

I'll add this to the point cloud tracker issue #536

**Comment by Gmadges** [comment]
All signed 👍

Regarding bounds calculation, would that be a good opportunity to use math.gl? I had a quick look at the api and there is some useful functions BoundingSphere & Oriented Bounding Box. It may be overkill for smaller meshes & point clouds but i could see some usefulness on larger point clouds.

**Comment by ibgreen** [comment]
> Regarding bounds calculation, would that be a good opportunity to use math.gl? I had a quick look at the api and there is some useful functions BoundingSphere & Oriented Bounding Box. It may be overkill for smaller meshes & point clouds but i could see some usefulness on larger point clouds.

You are looking at how to make the various vis.gl frameworks work better together, I like it!

A general aspiration behind the vis.gl framework suite is (within reasonable extents) to have frameworks that  work seamlessly together, without actually depending on each other. I.e. we want to:
-  avoid building one huge monolith where everything imports everything else and even the simplest app gets a huge bundle size.
- while making it trivial to import and compose exactly the functionality the app needs.

Also I make the following observations:
- The algorithms that calculate `BoundingSpheres` and `OrientedBoundingBoxes` are fairly complex and can thus represent undesirably large dependencies for what are supposed to be relatively small loader objects. 
- For many applications, AxisAligned bounding boxes are good enough, they are simple to use and do a reasonable job for just centering viewports.

I feel that a good compromise would be to have loaders provide a simple axis aligned bounding box (i.e. min/max for x/y/z like you currently do), while adding documentation to the category page for mesh loaders that mentions the availability of the more advanced math.gl methods

Note that the math.gl algorithms you mention, currently only work on "arrays of arrays" `[[x, y, z], [x, y, z]]` but not on inline attributes (i.e array of `x, y, z, x, y, z`). So they currently fail the "easily composable with loaders.gl" test. But that should be fixed in math.gl, not in loaders.gl.

With that in mind, the ideal approach would be:
- [ ] Add your simple bounding box calculations to every mesh loader.
- [ ] Document this in a mesh loader category doc page
- [ ] Also document there that more advanced bounding box calculations are available in math.gl
- [ ] Update the math.gl methods to work on "inline arrays`.

I could help with the last item, if you feel it is a bit out of scope, or provide detailed directions.

If this seems to ambitious / does not make sense let me know.

**Comment by Gmadges** [comment]
No this sounds great and makes sense to me. 👍 
I'll hopefully submit another PR with these changes over the next couple of weeks (if I'm not too busy with work).

**Comment by ibgreen** [comment]
Fully understood. 

Having regular / semi-regular contributors is of course extremely valuable to open source projects. So, your experience as contributor matters a lot! If you think this is a framework that inspires you, feel free to provide feedback on how we can facilitate participation for yourself and others.

**Pull request merged**